### PR TITLE
[v10.4.x] Chore: Improve domain validation for Google OAuth - Backport 83229 to v10.4.x

### PR DESCRIPTION
**Pull request by linoman**
**What is this feature?**

This PR improves domain validation between the response token id `hd` claim and the `allowed_domains` configuration setting for Google OAuth.

Backport from https://github.com/grafana/grafana/pull/83229

**Why do we need this feature?**

[Add a description of the problem the feature is trying to solve.]

**Who is this feature for?**

Anyone using Google OAuth.

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

# Release notice breaking change

We're adding a between the response of the ID token HD parameter and the list of allowed domains. This feature can be disabled through the configuration toggle `validate_hd `. Anyone using the legacy Google OAuth configuration should disable this validation if the ID Token response doesn't have the HD parameter.

### CODE REVIEW

File: docs/sources/setup-grafana/configure-security/configure-authentication/google/index.md [hunk 1]
```diff
@@ -111,6 +111,14 @@ automatically signed up.
 You may specify a domain to be passed as `hd` query parameter accepted by Google's
 OAuth 2.0 authentication API. Refer to Google's OAuth [documentation](https://developers.google.com/identity/openid-connect/openid-connect#hd-param).
 
+{{% admonition type="note" %}}
+Since Grafana 10.3.0, the `hd` parameter retrieved from Google ID token is also used to determine the user's hosted domain. The Google Oauth `allowed_domains` configuration option is used to restrict access to users from a specific domain. If the `allowed_domains` configuration option is set, the `hd` parameter from the Google ID token must match the `allowed_domains` configuration option. If the `hd` parameter from the Google ID token does not match the `allowed_domains` configuration option, the user is denied access.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/google/index.md:115]: @linoman , Same question here as with the 10.3 backport PR.

File: docs/sources/setup-grafana/configure-security/configure-authentication/google/index.md [hunk 2]
```diff
@@ -111,6 +111,14 @@ automatically signed up.
...
+
+When an account does not belong to a google workspace, the hd claim will not be available.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/google/index.md:117]:
```suggestion
When an account does not belong to a Google Workspace, the` hd` claim is not available.
```

**Review by chri2547 — COMMENTED** [review]
@linoman , similar feedback as with the 10.3 backport PR. Thanks!

**Comment by linoman** [comment]
@chri2547 Thank your for your comments; I've addressed them as requested.

File: pkg/login/social/connectors/google_oauth.go [hunk 3]
```diff
@@ -17,32 +17,38 @@ import (
...
 	Name          string `json:"name"`
 	EmailVerified bool   `json:"email_verified"`
+	HD            string `json:"hd"`
 	rawJSON       []byte `json:"-"`
 }
 
 func NewGoogleProvider(info *social.OAuthInfo, cfg *setting.Cfg, ssoSettings ssosettings.Service, features featuremgmt.FeatureToggles) *SocialGoogle {
 	provider := &SocialGoogle{
 		SocialBase: newSocialBase(social.GoogleProviderName, info, features, cfg),
+		validateHD: MustBool(info.Extra[validateHDKey], false),
```
Comment (IevaVasiljeva) [inline pkg/login/social/connectors/google_oauth.go:51]: So we want the default to be `true` or `false`? Asking, as the default used here is different from the default in the config file.

**Review by IevaVasiljeva — APPROVED** [review]
Approved with a small question about the default values.

File: pkg/login/social/connectors/google_oauth.go (hunk 3 above)
Comment (linoman) [reply pkg/login/social/connectors/google_oauth.go:51]: Good catch! This is set to true for Grafana 11.0, for this backports, the config value for `validate_hd` is `false`.

**Pull request merged**